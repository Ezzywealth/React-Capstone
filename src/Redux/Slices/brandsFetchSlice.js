import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  brands: [],
  phones: [],
  details: {},
  specs: {},
  loading: false,
};

export const fetchBrands = createAsyncThunk('fetchBrands', async () => {
  const { data } = await axios.get(
    'http://phone-specs-api.azharimm.dev/brands',
  );
  return data.data;
});

export const fetchPhones = createAsyncThunk('fetchPhones', async (slug) => {
  const { data } = await axios.get(
    `http://phone-specs-api.azharimm.dev/brands/${slug}`,
  );
  return data.data.phones;
});

export const fetchPhoneDetails = createAsyncThunk(
  'fetchPhoneDetails',
  async (slug) => {
    const { data } = await axios.get(
      `http://phone-specs-api.azharimm.dev/${slug}`,
    );
    return data.data;
  },
);

const brandSlice = createSlice({
  name: 'fetchBrands',
  initialState,
  reducers: {
    setPhoneSpecs: (state, action) => {
      const newState = { ...state };
      newState.specs = action.payload;
      return newState;
    },
    setLoadingFalse: (state) => {
      const newState = { ...state };
      newState.loading = false;
      return newState;
    },
    setLoadingTrue: (state) => {
      const newState = { ...state };
      newState.loading = false;
      return newState;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchBrands.pending, (state) => {
      const newState = { ...state };
      newState.loading = true;
      return newState;
    });
    builder.addCase(fetchBrands.fulfilled, (state, action) => {
      const newState = { ...state };
      newState.brands = action.payload;
      newState.loading = false;
      return newState;
    });
    builder.addCase(fetchPhones.pending, (state) => {
      const newState = { ...state };
      newState.loading = true;
      return newState;
    });
    builder.addCase(fetchPhones.fulfilled, (state, action) => {
      const newState = { ...state };
      newState.phones = action.payload;
      newState.loading = false;
      return newState;
    });
    builder.addCase(fetchPhoneDetails.pending, (state) => {
      const newState = { ...state };
      newState.loading = true;
      return newState;
    });
    builder.addCase(fetchPhoneDetails.fulfilled, (state, action) => {
      const newState = { ...state };
      newState.details = action.payload;
      newState.loading = false;
      return newState;
    });
  },
});

export default brandSlice.reducer;
export const { setPhoneSpecs, setLoadingFalse } = brandSlice.actions;
