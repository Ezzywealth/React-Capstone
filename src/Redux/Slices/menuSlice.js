import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  menuState: false,
  title: 'Home',
};

const menuSlice = createSlice({
  name: 'menuSlice',
  initialState,
  reducers: {
    setTitle: (state, action) => {
      const newState = { ...state };
      newState.title = action.payload;
      return newState;
    },
  },
});

export default menuSlice.reducer;
export const { setTitle } = menuSlice.actions;
