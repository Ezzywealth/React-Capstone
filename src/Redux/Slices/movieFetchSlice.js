import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  brands: [],
};

export const fetchBrands = createAsyncThunk('fetchBrands', async () => {
  const { data } = await axios.get(
    // 'http://phone-specs-api.azharimm.dev/infinix_hot_12-11498',
    // 'http://phone-specs-api.azharimm.dev/brands/infinix-phones-119',
    'http://phone-specs-api.azharimm.dev/brands',
  );
  return data.data;
});

const brandSlice = createSlice({
  name: 'fetchBrands',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchBrands.fulfilled, (state, action) => {
      const newState = { ...state };
      newState.brands = action.payload;
      return newState;
    });
  },
});

export default brandSlice.reducer;
