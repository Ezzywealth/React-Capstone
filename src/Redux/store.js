import { configureStore } from '@reduxjs/toolkit';
import fetchReducer from './Slices/brandsFetchSlice';
import menuReducer from './Slices/menuSlice';

const store = configureStore({
  reducer: {
    brands: fetchReducer,
    menu: menuReducer,
  },
});

export default store;
