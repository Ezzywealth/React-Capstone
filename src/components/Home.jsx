import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { fetchBrands } from '../Redux/Slices/brandsFetchSlice';
import { setTitle } from '../Redux/Slices/menuSlice';
import Brands from './Brands';

function Home() {
  const dispatch = useDispatch();
  const { brands } = useSelector((state) => state.brands);

  useEffect(() => {
    dispatch(fetchBrands());
    dispatch(setTitle('Brands'));
  }, []);

  const handleBrandClick = (link) => {
    dispatch(fetchBrands(link));
  };

  if (brands.length < 1) {
    return <h2>Loading........</h2>;
  }
  return (
    <main className="">
      <article className=" homeBg flex  flex-col text-white items-end p-8">
        <h3 className="flex flex-col text-2xl font-bold items-end">
          2023
          <span>Phone Specs</span>
        </h3>
        <h4 className="flex flex-col items-center">
          {`${brands.length}
          Brands`}
        </h4>
      </article>
      <h3 className="text-white px-3 py-1 border bg-rose-500 border-[#e0407d] text-xl shadow-xl">
        All Brands Stats
      </h3>
      <div className="flex w-full px-4 py-3 border-rose-500 border bg-rose-600 text-rose-200 font-semibold">
        <h3 className="flex flex-col text-xl flex-1 justify-between items-center">
          {brands.length}
          <span className="text-sm"> Brands</span>
        </h3>
        <h3 className="flex flex-col text-xl flex-1 justify-between items-center">
          {brands.length * 33}
          <span className="text-sm">Views</span>
        </h3>
        <h3 className="flex flex-col text-xl flex-1 justify-between items-center">
          {brands.length * 234}
          <span className="text-sm">Likes</span>
        </h3>
        <h3 className="flex flex-col text-xl flex-1 justify-between items-center">
          {brands.length * 223}
          <span className="text-sm">Orders</span>
        </h3>
      </div>
      <section className="cards grid grid-cols-2 w-full justify-center items-center">
        {brands.map((brand) => (
          <Link key={brand.brand_id} to={`/brands/${brand.brand_slug}`}>
            <section
              onClick={() => handleBrandClick(brand.details)}
              aria-hidden="true"
              className="border border-rose-400"
            >
              <Brands brand={brand} />
            </section>
          </Link>
        ))}
      </section>
    </main>
  );
}

export default Home;
