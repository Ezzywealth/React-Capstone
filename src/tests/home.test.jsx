import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import TestRenderer, { act } from 'react-test-renderer';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import Home from '../components/Home';
import store from '../Redux/store';
import { setLoadingFalse } from '../Redux/Slices/brandsFetchSlice';

describe('App', () => {
  it('renders', () => {
    // const history = createMemoryHistory();
    const tree = TestRenderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Home />
        </BrowserRouter>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Test for header', async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Home />
        </BrowserRouter>
      </Provider>,
    );

    act(() => {
      store.dispatch(setLoadingFalse());
    });
    const header = await screen.getByText(/All Brands Stats/i);
    expect(header).toBeInTheDocument();
  });
});
