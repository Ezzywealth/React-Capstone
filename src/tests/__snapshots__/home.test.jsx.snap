// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<main
  className=""
>
  <article
    className=" homeBg flex  flex-col text-white items-end p-8"
  >
    <h3
      className="flex flex-col text-2xl font-bold items-end"
    >
      2023
      <span>
        Phone Specs
      </span>
    </h3>
    <h4
      className="flex flex-col items-center"
    >
      0
          Brands
    </h4>
  </article>
  <h3
    className="text-white px-3 py-1 border bg-rose-500 border-[#e0407d] text-xl shadow-xl"
  >
    All Brands Stats
  </h3>
  <div
    className="flex w-full px-4 py-3 border-rose-500 border bg-rose-600 text-rose-200 font-semibold"
  >
    <h3
      className="flex flex-col text-xl flex-1 justify-between items-center"
    >
      0
      <span
        className="text-sm"
      >
         Brands
      </span>
    </h3>
    <h3
      className="flex flex-col text-xl flex-1 justify-between items-center"
    >
      0
      <span
        className="text-sm"
      >
        Views
      </span>
    </h3>
    <h3
      className="flex flex-col text-xl flex-1 justify-between items-center"
    >
      0
      <span
        className="text-sm"
      >
        Likes
      </span>
    </h3>
    <h3
      className="flex flex-col text-xl flex-1 justify-between items-center"
    >
      0
      <span
        className="text-sm"
      >
        Orders
      </span>
    </h3>
  </div>
  <section
    className="cards grid grid-cols-2 w-full justify-center items-center"
  />
</main>
`;
